centrallix/expression/exp_functions.c

```diff
@@ -4159,6 +4161,91 @@ int exp_fn_cos_compare(pExpression tree, pParamObjects objlist, pExpression i0,
+	return -1;
+	}
+    else if ((password->Flags | salt->Flags) & EXPR_F_NULL)
+	{
+	tree->Flags |= EXPR_F_NULL;
+	return 0;
+	}
+
+    // The default values of the following four variables should be tuned for each specific system's needs
+    // T_COST determines the number of passes the algorithm makes
+    unsigned int T_COST = ((tree->Children.nItems >= 3) && ((tree->Children.Items[2]) != NULL) && ((pExpression)tree->Children.Items[2])->DataType == DATA_T_INTEGER && ((pExpression)tree->Children.Items[2])->Integer < 8 && ((pExpression)tree->Children.Items[2])->Integer > 0) ? ((pExpression)tree->Children.Items[2])->Integer:2;
+    // M_COST determines the amount of memory cost in kilobytes
+    unsigned int M_COST = ((tree->Children.nItems >= 4) &&  ((tree->Children.Items[3]) != NULL) && ((pExpression)tree->Children.Items[3])->DataType == DATA_T_INTEGER && ((pExpression)tree->Children.Items[3])->Integer < INT_MAX && ((pExpression)tree->Children.Items[3])->Integer >= 64) ? ((pExpression)tree->Children.Items[3])->Integer:(1<<16);
+    // PARALLELISM determines the number of threads or 'lanes' used by the algorithm
+    unsigned int PARALLELISM = ((tree->Children.nItems) >= 5 && ((tree->Children.Items[4]) != NULL) && ((pExpression)tree->Children.Items[4])->DataType == DATA_T_INTEGER &&((pExpression)tree->Children.Items[4])->Integer <= 8 && ((pExpression)tree->Children.Items[4])->Integer >=1) ? ((pExpression)tree->Children.Items[4])->Integer:1;
+    // HASHLEN is the size of the finished hash
+    unsigned int HASHLEN = ((tree->Children.nItems >= 6) && ((tree->Children.Items[5]) != NULL) && ((pExpression)tree->Children.Items[5])->DataType == DATA_T_INTEGER && ((pExpression)tree->Children.Items[5])->Integer < 256 && ((pExpression)tree->Children.Items[5])->Integer >= 4) ? ((pExpression)tree->Children.Items[5])->Integer:32;
+    
+    // check if required parameters exist
+    if (!password || !salt)
+    	{
+	mssError(1, "EXP", "Invalid Parameters: function usage: exp_argon2id(password, salt)");
+	return -1;
+	}
+    tree->DataType = DATA_T_STRING;
+    
+    // hashvalue is where the output is written 
+    unsigned char hashvalue[HASHLEN];
+
+    unsigned char *slt = (unsigned char *)strdup(salt->String);
+    unsigned char *pwd = (unsigned char *)strdup(password->String);
+    unsigned int pwdlen = strlen((char*)pwd);
+    unsigned int sltlen = strlen((char*)slt);
+
+    // salt length check
+    if (sltlen < 8)
+	{
+	mssError(1, "EXP", "Salt is too short: Salt must be at least 8 bytes (16 recommended)");
+	return -1;
+	}
+    // this call to the argon2id_hash_raw method is where the magic happens
+    // after this call, the hashed password is written in hashvalue   
+    argon2id_hash_raw(T_COST, M_COST, PARALLELISM, pwd, pwdlen, slt, sltlen, hashvalue, HASHLEN);
+    
+    // this is where we write the contents of hashvalue to tree using qpfPrintf
+    if (HASHLEN*2+1 > 63)
+	{
```

Comment:
Instead of using the constant 63 here, use sizeof(tree->Types.StringBuf) - 1.  For example:
`if (HASHLEN*2+1 > sizeof(tree->Types.StringBuf))`
Since the nul terminator is already accounted for in the +1 on the left hand side, we don't have to subtract one from the length of the buffer.